KAFKA-2875:  remove slf4j multi binding warnings when running form source distribution

jinxing64 opened the pull request:
hi @ijuma I reopened this pr again (sorry for my inexperience using github);
I think I did much deduplication for the script;
Please have a look when you have time  : - )

Commit:
Merge pull request #12 from apache/trunk

2015-12-18

Commit:
KAFKA-2875 Class path contains multiple SLF4J bindings warnings when using scripts under bin

jinxing64 commented:
@ijuma  how do you think about this? Is it appropriate?

Commit:
Merge pull request #13 from apache/trunk

2015-12-30

ijuma commented on bin/kafka-console-consumer.sh:
```diff
@@ -18,4 +18,17 @@ if [ "x$KAFKA_HEAP_OPTS" = "x" ]; then
+        SCALA_BINARY_VERSION=2.10
+fi
+
+JARPATH="$base_dir/core/build/dependant-libs-${SCALA_VERSION}/*.jar \
```
Instead of configuring the JARPATH, could we not set something like:

PROJECT_NAMES=core,clients

And then use that to build the JARPATH in `kafka-run-class`?

ijuma replied:
That would mean that we can remove the boilerplate from this class.

ijuma commented on bin/kafka-run-class.sh:
```diff
@@ -31,65 +31,27 @@ if [ -z "$SCALA_BINARY_VERSION" ]; then
+  JARPATH="$base_dir/core/build/dependant-libs-${SCALA_VERSION}/*.jar \
+            $base_dir/examples/build/libs//kafka-examples*.jar \
+            $base_dir/contrib/hadoop-consumer/build/libs//kafka-hadoop-consumer*.jar \
+            $base_dir/contrib/hadoop-producer/build/libs//kafka-hadoop-producer*.jar \
```
The `/contrib/` folder was removed, so we can remove the two lines above.

ijuma commented:
Sorry for the delay @ZoneMayor, I left a comment. Please let me know what you think.

Commit:
Merge pull request #14 from apache/trunk

2016-1-10

Commit:
Merge branch 'trunk-KAFKA-2875' of https://github.com/ZoneMayor/kafka into trunk-KAFKA-2875

Commit:
KAFKA-2875: wip

Commit:
KAFKA-2875: Classpath contains multiple SLF4J bindings warnings when using scripts under bin

jinxing64 commented:
@ijuma 
thanks a lot for your comment;
It make sense to use 'PROJECT_NAMES';
I made some fix, please take a look when you have time;

ijuma commented on bin/kafka-run-class.sh:
```diff
@@ -30,66 +30,44 @@ if [ -z "$SCALA_BINARY_VERSION" ]; then
+      JARPATH=$JARPATH"$base_dir/core/build/dependant-libs-${SCALA_VERSION}/*.jar $base_dir/core/build/libs/kafka_${SCALA_BINARY_VERSION}*.jar "
+      ;;
+      "tools")
+      JARPATH=$JARPATH"$base_dir/tools/build/libs/kafka-tools*.jar $base_dir/tools/build/dependant-libs-${SCALA_VERSION}/*.jar "
```
The tools jar should not have a dependency on Scala code, right?

jinxing64 replied:
tools is currently for ProducerPerformance ProducerPerformance, VerifiableConsumer, VerifiableProducer, they have dependancies like (net.sourceforge.argparse4j:argparse4j:0.5.0)
in build.gradle;

``` java
    tasks.create(name: "copyDependantLibs", type: Copy) {
        from (configurations.testRuntime) {
            include('slf4j-log4j12*')
        }
        from (configurations.runtime) {
            exclude('kafka-clients*')
        }
        into "$buildDir/dependant-libs-${resolvedScalaVersion}"
    }
```

So, I think we need to keep `$base_dir/tools/build/dependant-libs-${SCALA_VERSION}/*.jar`

ijuma replied:
argparse4j is Java. And those tools are also in Java. I think the tools build is buggy with regards to the output folder. Can you please double-check?

granthenke replied:
Agreed this should probably be "$buildDir/dependant-libs" in the gradle build and then no scala version is needed in the path. 

I am also a bit confused why some projects use the dependant-libs directories, and others don't.

ijuma replied:
The projects that don't use it, do they have dependencies?

granthenke replied:
@ijuma Clients doesn't use it. Everything just happens to work because we get them from core.

ijuma commented:
@ZoneMayor Thanks, I left one question, the rest seems promising. There are some conflicts with trunk, can you please fix them?

granthenke commented on bin/connect-distributed.sh:
```diff
@@ -20,4 +20,4 @@ if [ "x$KAFKA_LOG4J_OPTS" = "x" ]; then
 fi
 
-exec $(dirname $0)/kafka-run-class.sh org.apache.kafka.connect.cli.ConnectDistributed "$@"
+PROJECT_NAMES="clients connect" exec $(dirname $0)/kafka-run-class.sh org.apache.kafka.connect.cli.ConnectDistributed "$@"
```
Does this need to be on the same line as the exec call?

jinxing64 replied:
Yes, because pass PROJECT_NAMES to kafka-run-class.sh, but don't want to export this variable to global scope

granthenke commented on bin/kafka-run-class.sh:
```diff
@@ -30,66 +30,44 @@ if [ -z "$SCALA_BINARY_VERSION" ]; then
-do
-  CLASSPATH=$CLASSPATH:$file
-done
+# run ./gradlew copyDependantLibs to get all dependant jars in a local dir
```
Should this move up above the JARPATH code, or be removed?

jinxing64 replied:
I will remove this;

Commit:
Merge pull request #15 from apache/trunk

2016-01-28

Commit:
KAFKA-2875: remove slf4j multi binding warnings when running form source distribution (merge to trunk and resolve conflict)

Commit:
added a new line

jinxing64 commented:
@ijuma @granthenke 
Thanks a lot for your comment;
I solved the conflict with trunk ,and modified some places;
Please check~
It's great if it can be merged;

ijuma commented:
Thanks @ZoneMayor, did you see this comment https://github.com/apache/kafka/pull/693/files#r51100904 ?

granthenke commented:
Do we also need to update the windows (.bat) scripts?

granthenke commented:
Thinking out loud a bit here. Hopefully this doesn't derail too much:

Recent changes in the build have ensured we have consistent artifact versions across the modules. Because our build has a lot of inter-project dependencies we duplicate a lot of jars pulled from the dependent-libs directories onto the classpath, not just the slf4j ones. 

Would it make sense to change the build to generate a central dependent-libs directory with all of the de-duplicated jars? That can be done in this patch, or I am happy to do it with a new subtask under KAFKA-2787.

ijuma commented:
@granthenke It is a good question. Personally, I'd prefer if each command would only get the relevant jars in the classpath. A central dependent-libs directory makes that a bit harder, especially with wildcard classpath entries. Having said that, whatever solution we come up with has some drawbacks so I'm not actually sure what's the right thing to do here.

@ewencp What are your thoughts on the solution proposed here versus a central dependent-libs directory?

granthenke commented:
@ijuma I agree it would be optimal to have the correct classpath without extra jars. 

I like the approach used in this patch to only use the classpath of the need projects/modules. However, I think we could do it at build time instead of runtime. That way the scripts wouldn't need to "hardcode" their project dependencies and we could leave that logic in the gradle build. The gradle build is where we define what dependencies are required. We could still deduplicate jars too.

It makes the build a little more complicated, but one way to do this while building would be:
- place artifacts in a central lib dir
- place dependencies in a central deps dir
- symlink all deps from the central dirs in the modules dependency dir
  - Not sure about windows here as I am less familiar with linking
- use the changes in this patch that define the project and load that projects classpath only
  - Only difference is you only define 1 project, not multiple.

granthenke commented:
@ZoneMayor I apologize for not bringing this up earlier. Thank you for your persistence on this patch.

ijuma commented:
@granthenke Yes, using the gradle build to extract project dependencies is definitely better. Not sure about symlinking though, is the optimisation worth dealing with the potential compatibility issues?

jinxing64 commented:
@granthenke 
Yes, I very agree with you;
-  Make a central dependant-libs directory to contain the deduplicated jars
-  Resolve the dependancies inside of gradle build, create symlink from the dependencies or jars (can from the central dependant-libs or other dependant projects) to project build dir, and execute script with a classpath containing just inside of local project build dir;

By above two,  scripts don't need to hardcode the dependencies any more;

It will be great if I can do above within this patch : ) , I will try to fix later

Commit:
KAFKA-2875: create deduplicated dependant-libs and use symlink to construct classpath

Commit:
small fix

Commit:
KAFKA-2875: modify classpath in windows bat

Commit:
mod server.properties for test

Commit:
KAFKA-2875: small fix

Commit:
KAFKA-2875: small fix

jinxing64 commented:
@ijuma @granthenke 

I modified this patch
- Create de-duplicated dependant libs located at kafka/build
- Separate the dependant libs into four parts: java-dependant-libs, scala-dependant-libs, java-test-dependant-libs, scala-test-dependant-libs
- Create the symlink from de-duplicate dependant libs to project's build directory; to run script, we have to sepecify the project name in order to construct classpath only with jars needed;
- Also modified script for windows

granthenke commented on bin/kafka-run-class.sh:
```diff
@@ -30,56 +30,50 @@ if [ -z "$SCALA_BINARY_VERSION" ]; then
-  CLASSPATH=$CLASSPATH:$dir/*
-done
+if [ -z "$PROJECT_NAMES" ]; then
+  PROJECT_NAMES="core tools clients connect-json connect-api connect-file connect-runtime stream log4j-appender"
```
If there is no PROJECT_NAME defined, perhaps just load all the jars in the central deps dir? Instead of all the projects. Then PROJECT_NAMES could be PROJECT_NAME and you wouldn't need a for loop below.

granthenke commented on build.gradle:
```diff
@@ -721,18 +861,33 @@ project(':connect:file') {
-  tasks.create(name: "copyDependantLibs", type: Copy) {
-    from (configurations.testRuntime) {
-      include('slf4j-log4j12*')
+  tasks.create(name: "copyDependantLibsJava", type: Copy) {
```
Since these tasks are the same (or should be) for all modules can they be de-duplicated and moved to the subprojects section of the build?

granthenke commented on build.gradle:
```diff
@@ -666,13 +789,15 @@ project(':connect:runtime') {
     compile libs.jettyServlet
     compile libs.reflections
 
+    runtime project(":connect:json")
```
Why did these need to be added?

Commit:
KAFKA-2875: replace PROJECT_NAMES with PROJECT_NAME, use the central deduplicated libs if PROJECT_NAME not specified

Commit:
small fix

jinxing64 commented:
@granthenke 
already modified the patch

`+    runtime project(":connect:json")`

Running Connect, we need connect:json to construct json converter;
Previously, the classpath include all the jars; 
Now we should take `connect:json` as a runtime dependency

jinxing64 commented:
@granthenke  @ijuma 
It's great if you can take a look at this : )

granthenke commented on bin/kafka-run-class.sh:
```diff
@@ -30,56 +30,47 @@ if [ -z "$SCALA_BINARY_VERSION" ]; then
+  JARPATH=$JARPATH" $base_dir/log4j-appender/build/libs/kafka-*.jar "
+  ;;
+  "")
+  JARPATH=$JARPATH" $base_dir/build/dependant-libs-java/*.jar"
```
Do we need the Scala dir here too?

granthenke commented on build.gradle:
```diff
@@ -318,16 +344,38 @@ project(':core') {
+      include('*scala*.jar')
     }
-    into "$buildDir/dependant-libs-${versions.scala}"
+    into "$baseBuildDir/dependant-libs-scala-${versions.scala}"
```
I think we can use `versions.baseScala` here. Will need to change the scripts to use SCALA_BINARY_VERSION too if you change this. But it results in `dependant-libs-scala-2.10` and  `dependant-libs-scala-2.11` instead of  `dependant-libs-scala-2.10.6` and  `dependant-libs-scala-2.11.7`

granthenke commented on build.gradle:
```diff
@@ -318,16 +344,38 @@ project(':core') {
+    from (configurations.testRuntime - configurations.runtime) {
+      include('*scala*.jar')
+    }
+    into "$baseBuildDir/test-dependant-libs-scala-${versions.scala}"
```
Same as above.

granthenke commented on build.gradle:
```diff
@@ -318,16 +344,38 @@ project(':core') {
+    into "$baseBuildDir/test-dependant-libs-scala-${versions.scala}"
+  }
+
+  task linkJars(dependsOn: ['copyDependantLibsJava', 'copyDependantLibsScala', 'copyTestDependantLibsJava', 'copyTestDependantLibsScala']) {
```
Can linkJars be broken into 2 tasks (linkJars and linkScalaJars). That way linkScalaJars can stay here, but linkJars can be moved up to subprojects and de-duplicated like copyDependantLibsJava was.

granthenke commented on bin/kafka-run-class.sh:
```diff
@@ -30,56 +30,47 @@ if [ -z "$SCALA_BINARY_VERSION" ]; then
+  JARPATH=$JARPATH" $base_dir/tools/build/dependant-libs-java/*.jar $base_dir/tools/build/libs/kafka-tools*.jar"
+  ;;
+  "clients")
+  JARPATH=$JARPATH" $base_dir/clients/build/libs/kafka-clients*.jar "
```
Missing dependant-libs

granthenke commented on bin/kafka-run-class.sh:
```diff
@@ -30,56 +30,47 @@ if [ -z "$SCALA_BINARY_VERSION" ]; then
+  JARPATH=$JARPATH" $base_dir/streams/build/dependant-libs-java/*.jar $base_dir/streams/build/libs/connect-*.jar"
+  ;;
+  "log4j-appender")
+  JARPATH=$JARPATH" $base_dir/log4j-appender/build/libs/kafka-*.jar "
```
Missing dependant-libs

granthenke commented on build.gradle:
```diff
@@ -494,18 +537,19 @@ project(':tools') {
-        into "$buildDir/dependant-libs-${versions.scala}"
-    }
+    task linkJars(dependsOn: ['copyDependantLibsJava', 'copyTestDependantLibsJava']) {
+      Files.createDirectories(Paths.get("$buildDir/dependant-libs-java"))
```
All of the work inside the linkJars task should be wrapped in `doLast { ... }` to prevent it from from executing at configuration time. 

You can see this break by running `gradle clean build -x test`. It will show a FileNotFound exception because at configuration time the copyDependantLibs tasks haven't run yet. Make sure you clean each time when testing changes to catch things like this.

granthenke commented on build.gradle:
```diff
@@ -318,16 +344,38 @@ project(':core') {
+    Files.createDirectories(Paths.get("$buildDir/test-dependant-libs-java"))
+    Files.createDirectories(Paths.get("$buildDir/test-dependant-libs-scala-${versions.scala}"))
 
+    copyDependantLibsJava.eachFile {
```
I don't think [eachFile](https://docs.gradle.org/current/dsl/org.gradle.api.tasks.Copy.html#org.gradle.api.tasks.Copy:eachFile%28groovy.lang.Closure%29) is what you want since it executes the task before its copied.

granthenke commented on build.gradle:
```diff
@@ -154,6 +158,22 @@ subprojects {
+  tasks.create(name: "copyDependantLibsJava", type: Copy) {
+    from(configurations.runtime) {
+      include('*.jar')
+      exclude('*scala*')
```
I am not sure if this is reliable. It may be tough to separate the scala and java jars. I guess it may work if you included the patter "_$versions. baseScala" which resolves to _2.10 or _2.11.

granthenke commented on bin/kafka-run-class.sh:
```diff
@@ -30,56 +30,47 @@ if [ -z "$SCALA_BINARY_VERSION" ]; then
-done
-
-# classpath addition for release
-CLASSPATH=$CLASSPATH:$base_dir/libs/*
```
The current logic in `releaseTarGz` was copying everything into a central libs directory here. If we are going to have a directory/classpath per project in both releases and source builds that would need to be updated too...

jinxing64 replied:
I followed current logic for releaseTarGz, added $baseDir/lib to be part of classpath;

Commit:
KAFKA-2875:wip

Commit:
KAFKA-2875: fixes

Commit:
KAFKA-2875: fix

jinxing64 commented:
@granthenke 
Thanks for you comment;
I modified this patch accordingly, please take a look : )

ijuma commented:
@ZoneMayor have you tested this on Windows?

jinxing64 commented:
@ijuma 
yes, I did the test on Windows;
It's great if you can leave some comment : )

Commit:
Merge pull request #16 from apache/trunk

2016-02-16

Commit:
KAFKA-2875: use slf4j 1.7.13

jinxing64 commented:
@ijuma  please take a look at this when you have time :)

ijuma commented:
Thanks @ZoneMayor, it seems reasonable, I need to spend some time testing it. @granthenke, any more thoughts?

jinxing64 commented:
@ijuma How about the testing result? : )

Commit:
Merge pull request #19 from apache/trunk

02-21

jinxing64 commented:
@ijuma  ping for review
How about the test : ) ?

Commit:
Merge pull request #20 from apache/trunk

2016-02-25

Commit:
Merge pull request #21 from apache/trunk

2016-03-11

Commit:
merge to trunk

jinxing64 commented:
@ijuma 
do you mind to take another look at this ?

guozhangwang commented:
I think it is great to have this resolved. @jinxing64 Would you like to rebase? If yes I could help merging it as is (cc @ijuma ).
